packages/services/src/boost-bonus.service.ts

```diff
@@ -517,6 +517,268 @@ export class BoostBonusService {
+          competitionId,
+          transaction,
+        );
+
+      if (boostChanges.length === 0) {
+        this.#logger.debug(
+          { competitionId },
+          "No bonus boosts found for competition - nothing to cleanup",
+        );
+        return {
+          removedBoostIds: [],
+          keptBoostIds: [],
+        };
+      }
+
+      this.#logger.info(
+        { competitionId, boostChangeCount: boostChanges.length },
+        "Found boost_changes entries for cleanup evaluation",
+      );
+
+      // Step 2: Get unique boost bonus IDs
+      const boostBonusIds = [
+        ...new Set(
+          boostChanges
+            .map((change) => change.meta?.boostBonusId as string | undefined)
+            .filter((id): id is string => id !== undefined),
+        ),
+      ];
+
+      if (boostBonusIds.length === 0) {
+        this.#logger.debug(
+          { competitionId },
+          "No boost_changes entries have boostBonusId in meta - nothing to cleanup",
+        );
+        return {
+          removedBoostIds: [],
+          keptBoostIds: [],
+        };
+      }
+
+      // Step 3: Fetch boost_bonus entries
+      const boosts = await this.#boostRepository.findBoostBonusesByIds(
+        boostBonusIds,
+        transaction,
+      );
+
+      this.#logger.info(
+        { competitionId, boostCount: boosts.length },
+        "Retrieved bonus boost entries for cleanup evaluation",
+      );
+
+      // Step 4: Process each boost - remove if invalid
+      for (const boost of boosts) {
+        const boostBonusId = boost.id;
+
+        // Check if boost is now invalid: newBoostStartDate >= boost.expiresAt
+        // Using >= because if they're equal, window opens when boost expires (invalid)
+        const isInvalid = newBoostStartDate >= boost.expiresAt;
+
+        if (!isInvalid) {
+          this.#logger.debug(
+            {
+              boostBonusId,
+              competitionId,
+              newBoostStartDate,
+              expiresAt: boost.expiresAt,
+            },
+            "Boost still valid after boostStartDate change - keeping",
+          );
+          keptBoostIds.push(boostBonusId);
+          continue;
+        }
+
+        // Boost is now invalid - remove it
+        this.#logger.info(
+          {
+            boostBonusId,
+            competitionId,
+            newBoostStartDate,
+            expiresAt: boost.expiresAt,
+          },
+          "Boost now invalid (boostStartDate >= expiresAt) - removing from competition",
+        );
+
+        // Get user to retrieve wallet address
+        const user = await this.#userRepository.findById(boost.userId);
+        if (!user) {
+          this.#logger.warn(
+            { boostBonusId, userId: boost.userId },
+            "User not found during boost cleanup - skipping",
+          );
+          continue;
+        }
+
+        // Create idempotency key for cleanup
+        const idemKey = TEXT_ENCODER.encode(
```

Comment:
If the same boost becomes invalid, gets cleaned up, then someone re-adds it and it becomes invalid again, this idempotency key would prevent the second cleanup, right? This is probably fine in practice, but worth documenting?

Reply:
documented